azurelinuxagent/ga/exthandlers.py

```diff
@@ -276,6 +288,73 @@ def run(self):
                       message=detailed_msg)
             return
 
+    def get_artifacts_profile(self):
+        try:
+            self.artifacts_profile = self.protocol.get_artifacts_profile()
+        except Exception as e:
+            msg = u"Exception retrieving VmArtifactsProfile blob: {0}".format(ustr(e))
+            detailed_msg = '{0} {1}'.format(msg, traceback.extract_tb(get_traceback(e)))
+            logger.warn(msg)
+
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.ExtensionProcessing,
+                      is_success=False,
+                      message=detailed_msg)
+
+    def process_goal_state(self, instantiation, sequence_number):
+        fabric_goal_state_changed, fast_track_changed = self.determine_what_changed(instantiation, sequence_number)
+        if fabric_goal_state_changed:
+            self.protocol.set_fast_track(False)
+
+        # Fabric goal state takes precedence over VMArtifactsProfile blob goal state
+        # So, process the Fabric goal state first
+        self.handle_ext_handlers(fabric_goal_state_changed)
+        self.last_instantiation = instantiation
+
+        # Now process the VMArtifactsProfile if it changed. Note that just because the VMArtifactsProfile changed
+        # doesn't mean the FastTrack goal state changed
+        if fast_track_changed and self.artifacts_profile is not None:
+            self.process_vm_artifacts_profile_for_fast_track(sequence_number)
+        elif self.protocol.client.get_use_fast_track():
+            # Our last goal state was via VMArtifactsProfile, so reset the extension handlers
+            self.ext_conf = self.protocol.client.get_ext_conf(force=True)
+            self.ext_handlers = self.ext_conf.ext_handlers
+
+    def determine_what_changed(self, instantiation, sequence_number):
+        fabric_goal_state_changed = False
+        fast_track_changed = False
+
+        if self.last_instantiation != instantiation:
+            logger.info('Goal state changed with instantiation={0}(last:{1})'.
+                        format(str(instantiation), str(self.last_instantiation)))
+            fabric_goal_state_changed = True
+
+        if sequence_number is not None and self.last_vm_artifacts_seqno != sequence_number and \
+                conf.get_extensions_fast_track_enabled():
+            logger.info('VMArtifactsProfile changed with sequence number={0}(last:{1})'.
+                        format(str(sequence_number), str(self.last_vm_artifacts_seqno)))
+            fast_track_changed = True
+        return fabric_goal_state_changed, fast_track_changed
+
+    def process_vm_artifacts_profile_for_fast_track(self, sequence_number):
+        vm_artifacts_ext_conf = self.artifacts_profile.transform_to_extensions_config()
```

Comment:
You can move this transformation earlier in the call stack.

Reply:
ok, moved up one